pkg/nodeagent/controller/operatingsystemconfig/containerd.go

```diff
@@ -0,0 +1,363 @@
+	for _, registryConfig := range registriesWithReadiness {
+		fns = append(fns, func(ctx context.Context) error {
+			return addRegistryToContainerdFunc(ctx, log, registryConfig, r.FS)
+		})
+	}
+
+	go func() {
+		errChan <- flow.Parallel(fns...)(ctx)
+	}()
+
+	return errChan
+}
+
+func addRegistryToContainerdFunc(ctx context.Context, log logr.Logger, registryConfig extensionsv1alpha1.RegistryConfig, fs afero.Afero) error {
+	httpClient := http.Client{Timeout: 1 * time.Second}
+
+	baseDir := path.Join(certsDir, registryConfig.Upstream)
+	if err := fs.MkdirAll(baseDir, defaultDirPermissions); err != nil {
+		return fmt.Errorf("unable to ensure registry config base directory: %w", err)
+	}
+
+	hostsTomlFilePath := path.Join(baseDir, "hosts.toml")
+	exists, err := fs.Exists(hostsTomlFilePath)
+	if err != nil {
+		return fmt.Errorf("unable to check if registry config file exists: %w", err)
+	}
+
+	// Check if registry endpoints are reachable if the config is new.
+	// This is especially required when registries run within the cluster and during bootstrap,
+	// the Kubernetes deployments are not ready yet.
+	if !exists && ptr.Deref(registryConfig.ReadinessProbe, false) {
+		log.Info("Probing endpoints for image registry", "upstream", registryConfig.Upstream)
+		if err := retry.Until(ctx, 2*time.Second, func(ctx context.Context) (done bool, err error) {
+			for _, registryHosts := range registryConfig.Hosts {
+				req, err := http.NewRequestWithContext(ctx, http.MethodGet, registryHosts.URL, nil)
+				if err != nil {
+					return false, fmt.Errorf("failed to construct http request %s for upstream %s: %w", registryHosts.URL, registryConfig.Upstream, err)
+				}
+
+				_, err = httpClient.Do(req)
+				if err != nil {
+					return false, fmt.Errorf("failed to reach registry %s for upstream %s: %w", registryHosts.URL, registryConfig.Upstream, err)
+				}
+			}
+			return true, nil
+		}); err != nil {
+			return err
+		}
+
+		log.Info("Probing endpoints for image registry succeeded", "upstream", registryConfig.Upstream)
+	}
+
+	f, err := fs.OpenFile(hostsTomlFilePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("unable to open hosts.toml: %w", err)
+	}
+
+	defer func() {
+		err = f.Close()
+	}()
+
+	type (
+		hostConfig struct {
+			Capabilities []extensionsv1alpha1.RegistryCapability `toml:"capabilities,omitempty"`
+			CaCerts      []string                                `toml:"ca,omitempty"`
+		}
+
+		config struct {
+			Server *string               `toml:"server,omitempty" comment:"managed by gardener-node-agent"`
+			Host   map[string]hostConfig `toml:"host,omitempty"`
+		}
+	)
+
+	content := config{
+		Server: registryConfig.Server,
+		Host:   map[string]hostConfig{},
+	}
+
+	for _, host := range registryConfig.Hosts {
+		h := hostConfig{}
+
+		if len(host.Capabilities) > 0 {
+			h.Capabilities = host.Capabilities
+		}
+		if len(host.CACerts) > 0 {
+			h.CaCerts = host.CACerts
+		}
+
+		content.Host[host.URL] = h
```

Comment:
Is the order of the hosts preserved during serialization? I am asking because the order of the host sections is respected by containerd. From https://github.com/containerd/containerd/blob/cbb2fc78e93afa5f68caf6780169a07d55d698a1/docs/hosts.md?plain=1#L228:
> When `host`(s) are specified, the hosts will be tried first in the order listed.

This might be relevant for the registry-mirror case in the registry-cache extension where we allow multiple hosts to be specified.

Reply:
Very nice catch! Looking at the marshaling logic, I think the order is **not preserved**. Let me check a few options and get back.

Reply:
I revised the procedure in 34924396c691f26bf317ad78f77ff0ab1731e475.